### PR TITLE
fix static linkage and make THD statically linked

### PR DESCRIPTION
**Pull request by soumith**
port over commit from v0.2.0 https://github.com/pytorch/pytorch/commit/0b92e5c9ed1b62e695b10167f87621bbbcf0fc86

### CODE REVIEW

**Comment by soumith** [comment]
looks like introducing broadcasting in ATen broke one of the JIT tests. cc: @gchanan @killeent @ezyang 

```
Running JIT tests
.........E..............
======================================================================
ERROR: test_cpp (__main__.TestJit)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test_jit.py", line 372, in test_cpp
    torch._C._jit_run_cpp_tests()
RuntimeError: The size of tensor a (32) must match the size of tensor b (128) at non-singleton dimension 2

----------------------------------------------------------------------
Ran 24 tests in 5.456s

FAILED (errors=1)
```

**Comment by ezyang** [comment]
@zdevito wrote this test case

File: setup.py [hunk 1]
```diff
@@ -406,7 +397,7 @@ def run(self):
         ]
         extra_compile_args += ['-DWITH_DISTRIBUTED_MW']
     include_dirs += [tmp_install_path + "/include/THD"]
-    main_link_args += [THD_LIB]
+    main_link_args += [THD_LIB, GLOO_LIB]
```
Comment (apaszke) [inline setup.py]: Don't you have to guard this with if Linux?

**Comment by ezyang** [comment]
Trevor and I think that the test case was always buggy, and the change for ATen to use new broadcasting just exposed the problem. I'm verifying this right now.

**Pull request merged**